hashintel/hash
# Allow disabling of user/org creation in the HASH workspace

## 🌟 What is the purpose of this PR?




This PR makes it possible to disable user and org creation in the HASH App, by introducing two new properties on the HASH instance entity type:
1. `userSelfRegistrationIsEnabled` - whether or not a user can self register (i.e. sign-up) with HASH.
1. `userRegistrationByInviteIsEnabled` - whether or not an existing user is able to register another user by inviting them to an org. Note this is not enforced as we don't have an org invite flow.
1. `orgSelfRegistrationIsEnabled` - whether or not a user can self-register an org (note this does not apply to instance admins). Note this is not enforced as we don't have a org registration flow for users.

When both (1) and (2) are `false`, user accounts can only be created by an instance administrator's action.

When only (1) is `false`, existing users will still be able to invite other users to a HASH instance (via -- when supported -- sharing an entity/type with a specific email address, or by adding them to an org).

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203301853897838/1203249971975780/f) _(internal)_

## 🔍 What does this change?




See commits.

## 📜 Does this require a change to the docs?



No.

## ⚠️ Known issues



## 🐾 Next steps



## 🛡 What tests cover this?



None.

## ❓ How to test this?



1. Run HASH using `yarn dev`
3. Observe that the sign up button is being displayed
4. Login as any user
5. Use the `updateEntityWithMetadata` GQL query to set `userSelfRegistrationIsEnabled` to `false`
```GraphQL
mutation updateEntityWithMetadata($updatedProperties:JSONObject!) {
  updateEntityWithMetadata(
    entityId:"00000000-0000-0000-0000-000000000000%<hashInstanceUuid>",
    updatedProperties: $updatedProperties
  ) 
}
```

```JSON
{
 "updatedProperties": {
  "http://localhost:3000/@example-org/types/property-type/user-self-registration-is-enabled/": true,
  ...remainingProperties,
}
```
6. Sign out, and observe that the sign up button disappears from the UI.
